### PR TITLE
Group Shared Folders

### PR DESCRIPTION
**Pull request by dkocher**

### CODE REVIEW

File: deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java [hunk 1]
```diff
@@ -90,6 +100,41 @@ public String getThirdLevelId(final Path file) throws BackgroundException {
...
+        if(containerService.isSharedWithMe(file)) {
+            return file;
+        }
+        final Deque<Path> segments = this.decompose(file);
+        Path result = new Path(String.valueOf(Path.DELIMITER), EnumSet.of(Path.Type.volume, Path.Type.directory));
+
+        while(!segments.isEmpty()) {
+            Path segment = segments.pop();
+            EnumSet<AbstractPath.Type> type = EnumSet.copyOf(segment.getType());
+            type.add(AbstractPath.Type.shared);
```
Comment (dkocher) [inline deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java:116]: Should only be added for files in _Shared_ container?

File: deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java [hunk 2]
```diff
@@ -82,14 +93,53 @@ public String getBoxNodeId(final Path file) throws BackgroundException {
...
+                if(matcher.matches()) {
+                    final String deepboxName = matcher.group(1);
+                    final String boxName = matcher.group(2);
+                    final EnumSet<AbstractPath.Type> type = EnumSet.copyOf(segment.getType());
+                    type.add(AbstractPath.Type.shared);
+                    final Path deepbox = new Path(result, deepboxName, type, new PathAttributes(segment.attributes()).withFileId(null));
+                    result = new Path(deepbox, boxName, type, segment.attributes());
+                }
+                else {
+                    log.warn("Could not match pattern");
```
Comment (dkocher) [inline deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java:128]: Improve logging to include name that did not match pattern.

File: deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxListService.java [hunk 3]
```diff
@@ -285,6 +323,40 @@ public AttributedList<Path> list(final Path directory, final ListProgressListene
                     offset += chunksize;
                 }
                 while(offset < size);
+                final Path shared = new Path(directory, containerService.getPinnedLocalization(SHARED), EnumSet.of(Path.Type.directory, Path.Type.volume));
+                if(!new SharedWithMeListService(companyId).list(shared, listener).isEmpty()) {
```
Comment (dkocher) [inline deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxListService.java:327]: Is it a requirement that the _Shard with me_ folder is only displayed conditionally?

File: deepbox/src/test/java/ch/cyberduck/core/deepbox/DeepboxListServiceTest.java [hunk 4]
```diff
@@ -117,6 +137,27 @@ public void testListBoxes() throws Exception {
...
+        for(final Path f : list) {
+            assertSame(shared, f.getParent());
+            assertFalse(f.getName().contains(String.valueOf(Path.DELIMITER)));
+            // no modification/creation date for Boxes
+            assertTrue(f.attributes().getModificationDate() < 0);
+            assertTrue(f.attributes().getCreationDate() < 0);
+            assertNotNull(nodeid.getFileId(new Path(f).withAttributes(new PathAttributes())));
+            assertEquals(f.attributes(), new DeepboxAttributesFinderFeature(session, nodeid).find(new Path(f.getAbsolute(), f.getType())));
+        }
+        session.close();
```
Comment (dkocher) [inline deepbox/src/test/java/ch/cyberduck/core/deepbox/DeepboxListServiceTest.java:158]: Already in `@After` of superclass.

File: deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxListService.java (hunk 3 above)
Comment (ylangisc) [reply deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxListService.java:327]: Yes.

File: deepbox/src/test/java/ch/cyberduck/core/deepbox/DeepboxListServiceTest.java (hunk 4 above)
Comment (ylangisc) [reply deepbox/src/test/java/ch/cyberduck/core/deepbox/DeepboxListServiceTest.java]: In [212d0c7](https://github.com/iterate-ch/cyberduck/pull/16420/commits/212d0c7693cfd27a369d4a980059b27e2da79ab3).

File: deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java (hunk 2 above)
Comment (ylangisc) [reply deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java]: In [212d0c7](https://github.com/iterate-ch/cyberduck/pull/16420/commits/212d0c7693cfd27a369d4a980059b27e2da79ab3).

**Review by ylangisc — APPROVED** [review]

**Pull request merged**